File: Java/basic/FirstRepeat.java
/*
 * This is free to use as it was only made for practice.
 */

package com.jhazelle.gfg.basic;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

/**
 * This is a solution to the <a href="http://practice.geeksforgeeks.org/problems/find-first-repeated-character/0" target="_blank">GeeksForGeeks Basic First Repeated Character.</a>
 * 
 * To summarize the practice:
 * Given a string, find the first character that repeats.
 * Print the repeated char if found, -1 if not.
 * This is similar to "FirstRepeatedCharacter" but this the goal here is to find the first occurance of a repeated character.
 * Note that in the string "geeksforgeeks" 'e' is the first repeated character, not 'g'.
 * This is because 'e' repeats before 'g' does.
 * 
 * @author <a href="mailto:jhazelle@asu.edu">Justin Hazelle</a>
 * <a href="https://github.com/BinaryWrought" target="_blank">GitHub</a>
 */
public class FirstRepeat 
{

    /**
     * @param args the command line arguments
     */
    public static void main( String[] args ) 
	{
		Scanner scanner = new Scanner( System.in );													//create a scanner object to get user input
		try																							//this try block is to catch the inevitable exception caused by the user quitting 
		{
			int t = scanner.nextInt();																//get number of test cases
			
			while( t > 0 )																			//So long as there are more test cases
			{				
				String s = scanner.next();															//get the string to be parsed
				Map< Character, Integer > map = new HashMap<>();									//container for the characters
				boolean found = false;																//flag to keep track of found repeated char
				
				for( int i = 0; i < s.length(); i++ )												//loop through all the chars in the string
				{
					if( !map.containsKey( s.charAt( i ) ) )											//check if the map already has this char
					{
						map.put( s.charAt( i ), 0 );												//if not, add it to the map
					}
					else
					{
						System.out.println( s.charAt( i ) );										//the char was already found in the map, so print it
						found = true;																//set the flag since a char was found repeated
						break;																		//break the loop as the first char was already found
					}

				}
				
				if( !found )																		//no repeating char was found
					System.out.println( "-1" );														//print -1
				
				t--;																				//get next test case		
			}
		}
		catch( Exception e )																		//catch an exception thrown by incorrect input
		{
			
		}
	}
}

File: Python/basic/FirstRepeat.py
#
# This is a solution to the <a href="http://practice.geeksforgeeks.org/problems/find-first-repeated-character/0" target="_blank">GeeksForGeeks Basic First Repeated Character.</a>
#
# To summarize the practice:
# Given a string, find the first character that repeats.
# Print the repeated char if found, -1 if not.
# This is similar to "FirstRepeatedCharacter" but this the goal here is to find the first occurance of a repeated character.
# Note that in the string "geeksforgeeks" 'e' is the first repeated character, not 'g'.
# This is because 'e' repeats before 'g' does.
#
# @author <a href="mailto:jhazelle@asu.edu">Justin Hazelle</a>
# <a href="https://github.com/BinaryWrought" target="_blank">GitHub</a>

def first_repeat(s):
    repeat_index = len(s)+1                 #container for the index of the repeating char
    repeat_char = "-1"                      #container for the char found to repeat
    for i in range(len(s)):                 #loop through each char in the string by index
        ri = s.find(s[i], i+1)              #find the index where this char repeats again
        if ri > -1 and ri < repeat_index:   #check if the char found at this index is found somewhere else in the string...
            repeat_char = s[ri]             #char was found to repeat, set this char as a candidate for first repeat
            repeat_index = ri               #char was found, remember the index for comparison
        if repeat_index <= i:               #a repeat was already found below the current index, no need to search the whole string
            break
    return repeat_char                      #return the repeated char, will be -1 if no repeats were found

t = int( input() )                          #get number of test cases
for _ in range( t ):                        #loop for each test case
    s = str( input() )                      #get the input for this test case
    print(first_repeat(s))                  #print the first repeating char
